### ISSUE TITLE
jackson-databind 2.15.0-rc1 introduced a subtle change on when BeanDeserializerModifier is invoked.

### ISSUE THREAD

**Issue by fdutton** [FasterXML/jackson-databind#4377]
### Describe your Issue

A few years ago I wrote a module to register a custom BeanDeserializerModifier that ensures collection-like properties are never null. The intent of using a BeanDeserializerModifier rather than registering specific deserializers was to honor any annotations on the property and any other deserializers registered for the property's type.
```
setDeserializerModifier(new BeanDeserializerModifier() {

    @Override
    public JsonDeserializer<?> modifyCollectionDeserializer(DeserializationConfig config, CollectionType type, BeanDescription beanDesc, JsonDeserializer<?> deserializer) {
        if (deserializer instanceof StringCollectionDeserializer original) {
            return new RequiredCollectionDeserializer(original, type);
        }
        if (deserializer instanceof CollectionDeserializer original) {
            return new RequiredCollectionDeserializer(original, type);
        }
        return deserializer;
    }

    @Override
    public JsonDeserializer<?> modifyMapDeserializer(DeserializationConfig config, MapType type, BeanDescription beanDesc, JsonDeserializer<?> deserializer) {
        if (deserializer instanceof MapDeserializer original) {
            return new RequiredMapDeserializer(original, type);
        }
        return deserializer;
    }

});
```

Starting with 2.15.0-rc1, these methods are not invoked unless I also change the behavior of ObjectMapper like so.
`mapper.setDefaultSetterInfo(JsonSetter.Value.forValueNulls(Nulls.AS_EMPTY));`

However, this seems to break constructor binding on regular POJOs. I get an InvalidDefinitionException.
`Cannot create empty instance of [simple type, class Foo], no default Creator`

Foo uses constructor binding and has only one constructor. I have tried annotating this constructor with `@JsonCreator` to no avail. It seems that Jackson is attempting to use setter binding.

I'm looking for some guidance on how to achieve the same result without asking for a Jackson change. I attempted to identify the actual change that introduced the new behavior but there are 6889 commits in 2.15.0-rc1 so I thought it would be easier to just ask first.

```
2.14.3     works
2.15.0-rc1 does not
2.16.1     does not
```

**Comment by cowtowncoder** [comment]
@fdutton It seems to me like this could be a clear flaw -- if description is accurate, those calls really should occur. Would it be easy enough to write a unit test just showing that no calls are made? Depending on what I'd find, I would try to backport a fix.
I would like to fix this problem, regardless of whether there are work-arounds.

As to work-arounds... there should be a way to limit application of `JsonSetter.Value.forValueNulls(Nulls.AS_EMPTY)` to just some types, using "config overrides" mechanism. Let me see if I can dig up an example.

**Comment by cowtowncoder** [comment]
Ok, so the general approach would be like (copied from a unit test):

```
        mapper.configOverride(String.class)
            .setSetterInfo(JsonSetter.Value.forValueNulls(Nulls.FAIL));
```

so you will specify target class on which to add `JsonSetter.Value` like `Nulls.AS_EMPTY`. That works for specific types; the one thing I am not 100% sure is whether it'd work given target type of `Collection.class` and/or `Map.class`.
Worth a try, I think, like:

```
        mapper.configOverride(Collection.class)
            .setSetterInfo(JsonSetter.Value.forValueNulls(Nulls.AS_EMPTY));
```

**Comment by fdutton** [comment]
@cowtowncoder Sorry, my description of the problem is incorrect. I tracked down the problem to the changes made for #3748.

When `getAbsentValue` was added in 2.13, it simply called `this.getNullValue(ctxt)`. Since DelegatingDeserializer did not override this method, the call chain would be getAbsentValue -> this.getNullValue -> delegate.getNullValue. My RequiredCollectionDeserializer class extends DelegatingDeserializer and overrides getNullValue and this worked because getAbsentValue invoked getNullValue in the same class.

The new call chain is getAbsentValue -> delegate.getAbsentValue -> this.getNullValue so my implementation of getNullValue never gets invoked. I just have to override getAbsentValue to fix the issue.
